man/man1/nemovar.1
.TH NEMOVAR 1NEMO "15 February 2025"

.SH "NAME"
nemovar \- get and set NEMOVAR variables

.SH "SYNOPSIS"
\fBnemovar\fP [parameter=value]

.SH "DESCRIPTION"
\fBnemovar\fP maintains a NEMO key/val storage server.  It will providing another method to
communicate (textual) information between tasks using key/val pairs.
These are persistent variables, and 
can cross different sessions, much like SHELL variables.
.PP
The current implementation uses a simple text file, but future implementation may also be able to
switch to shared memory, or whatever the operating system and environment provides, or what
the users' workflow prefers.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBvar=\fP
Name of the variable.
.br
Default: not used.
.TP
\fBval=\fP
If used, the variable is getting a new value.
.br
Default: not used.
.TP
\fB**delete=f|t\fP
Delete the named variable in \fBvar=\fP.
.br
**\fInot implemented yet\fP
.TP
\fB**load=\fP
Load from a file. Need a replace/merge option?
.br
**\fInot implemented yet\fP
.TP
\fB**save=\fP
Save to a file.
.br
**\fInot implemented yet\fP
.SH "EXAMPLES"
The \fBNEMOVAR\fP environment variable is used to set the interprocess communication, though
currently only files are supported.
For example on linux we might be able to use the shared memory as follows:
.EX

    export NEMOVAR=/dev/shm/nemovar.$USER
    nemovar a 10
    
.EE
to set the value of 10 to a. Then
.EX

    nemovar a
    
.EE
retrieves the value of a. And finally
.EX

    nemovar 
    
.EE
just shows all variables. Output is always to \fIstdout\fP. 

.SH "FORMAT"
Currently the format of the text file is a simple ASCII file, where all named variables are explictly shown as strings:
.EX

    a="10"
    b="hello world"
    c="3.141592"
    
.EE
to make it compatible with possible future \fIbash(1)\fP and \fIpython(1)\fP clients.
The data type interpretation is left to the client. This also opens the option for small lists, e.g.
.EX

    d="1,2,10:20:2"

.EE
which again leaves the client to parse the string, much like \fIgetparam(3NEMO)\fP. Here is the
way it can be done using \fInemoinp(1NEMO)\fP
.EX
    nemoinp $(nemovar d)
.EE

.SH "BACKGROUND"
There are several ways already to pass small textual information (outside of the large simulations files)
between NEMO programs, usually aided by the shell.

.PP
First of all, shell variables can be an effective way to control a script, but sometimes they need to
be extracted from a program, e.g. with some awk/grep/sed, or in this example with \fItxtpar(1NEMO)\fP
.EX
      ccdstat foo.ccd > foo.log
      mean=$(txtpar foo.log p0=Mean,1,5)
      disp=$(txtpar foo.log p0=Mean,1,6)
      ccdplot foo.ccd $mean-$disp,$mean,$mean+$disp
.EE
here we extracted the mean and dispersion from a map, and plotted contours of the mean, and one dispersion around the mean.
Assuming these variables from the previous example happen to be in a file called \fBnemopars.rc\fP, the following will
work
.EX
      mean=$(nemopars mean)
      disp=$(nemopars disp)
.EE
.PP
Simulations can store results in a \fIkeyword=value\fP textual database, something that can be easily read
by the shell, or python, or \fInemopars(1NEMO)\fP.
.PP
With the first version of nemovar you can do the following:
.EX
      export NEMOVAR=nemopars.rc
      mean=$(nemovar mean)
      disp=$(nemovar disp)
.EE

.PP
The power of nemovar will come when the user interface shortcuts the use of the shell, as in this
hypothetical example
.EX
     ccdstat foo.ccd nemovar=mean,disp
     ccdplot foo.ccd %mean-%disp,%mean,%mean+%disp'
.EE

.SH "FUTURE"
1. A future implementation will allow direct retrieval of NEMO variables is possible via the command line
(see \fIgetparam(3NEMO)\fP). For example, compare the two styles, \fInemovar\fP on the left,
shell variables on the right:
.EX

     nemovar eps 0.05                 eps=0.05
     hackcode1 eps=%eps               hackcode1 eps=$eps
     
.EE
Note that some \fItable(5NEMO)\fP
programs use %0, %1, %2, .... to refer to column numbers, so variable names cannot start with
a digit. Upper and Lower case should be allowed, and be different.
.PP
Recall that the current command line interface can already refer to other keywords, e.g.
.EX

    foobar a=1 b='$a'
    
.EE
would assign the same value to both a and b.

2. An option should exist to remove a variable from the database.

3. In case there is true shared memory not associated with a filename, there needs to be a way to save and load
true files.

4. When \fIoutkeys/savevars\fP are implemented, the following style of IPC is possible:

.EX

     ccdstat ccd1 outkeys=mean,rms
     ccdplot ccd1 %mean
    
.EE
plotting a single contour of the mean value in the image.

.SH "CAVEAT"
It is not possible to use nemovar's in a pipe, e.g. this is NOT reliable:
.EX
    program1 - outkeys=bar | program2 - foo=%bar
.EE


.SH "SEE ALSO"
txtpar(1NEMO), nemopars(8NEMO), getparam(3NEMO), table(3NEMO)
.PP
redis/valkey  https://github.com/valkey-io/valkey

.SH "FILES"
$NEMOVAR - currently the filename, but could be used to set a more generic IPC method.

.SH "AUTHOR"
Peter Teuben

.SH "HISTORY"
.nf
.ta +1.5i +5.5i
7-mar-2024	Created for NEMO2024	PJT
24-jun-2024	0.4:    write dummy if NEMOVAR does not exist	PJT
15-feb-2025	0.6:	remove quotes when single variables requested	PJT
.fi


src/kernel/io/nemovar.c
/*
 *  NEMOVAR:   display a NEMO variable from the $NEMOVAR file (or future shared memory)
 */

#include <nemo.h>
#include <table.h>
#include <extstring.h>
#include <filefn.h>
#include <sys/file.h>     // for flock(2) - alternatively flock(2) for NFS files

string defv[] = {
  "var=\n             Name of  variable(s)",
  "val=\n             Optional new value",
  "comment=\n         Optional comment when new value was set",
  "VERSION=0.6\n      15-feb-2025 PJT",
  NULL,
};

string usage="display NEMOVAR variable(s)";


void nemo_main()
{
  int i;
  string var = getparam("var");
  bool hasVar = hasvalue("var");
  string value = getparam("val");  
  bool hasVal = hasvalue("val");
  string comment = getparam("comment");
  bool hasComment = hasvalue("comment");  
  string nemovar = getenv("NEMOVAR");
  if (nemovar == NULL)
    error("$NEMOVAR was not set. Example: export NEMOVAR=/dev/shm/nemovar.$USER");
  dprintf(1,"nemovar: %s\n", nemovar);

  if (hasComment)
    warning("comment %s not yet used", comment);

  // bug or feature, nemovar currently needs to exist
  if (!fexist(nemovar)) {
    warning("Creating dummy %s", nemovar);
    stream ostr = stropen(nemovar,"w");
    fprintf(ostr,"# nemovar\n");
    strclose(ostr);
  }
  
  // read old nemovar
  stream tstr = stropen(nemovar,"r");
  table *t = table_open(tstr, -1);      // read table as a dumb set of lines
  int nrows = t->nr; 
  string s, sv, si, vareq;
  
  // check if just to show all variables
  if (!hasVar && !hasVal) {
    dprintf(1,"Show all variables\n");
    for (i=0; i<nrows; i++)
      printf("%s\n", table_row(t, i));
    return;
  }

  //@TODO - Should use hash in the future?
  // ensure we find the "var=" string
  vareq = allocate(strlen(var) + 2);
  sprintf(vareq,"%s=", var);
  dprintf(1,"vareq:%s\n",vareq);

  // @todo - this is where a delete= option would come out
  if (hasVar && !hasVal) { 
    // loop over lines in reverse and find the last keyword, then display it
    for (i=nrows-1; i>=0; i--) {
      s = table_row(t, i);
      if (strncmp(s,vareq,strlen(vareq))==0) {
	sv = s+strlen(var)+1;
	if (*sv == '"') {
	  sv++;
	  si=sv;
	  while (*si != '"')
	    si++;
	  *si=0;
	}
	printf("%s\n",sv);
	return;
      }    
    }
    dprintf(0,"Variable %s not found\n",var);
    return;
  }

  // writing code might need flock() to lock file
  // @todo currently this allows a variable to appear multiple times
  //       but all var's will get the new value
  //       'nemovar | sort | uniq' would solve that

  if (hasVar && hasVal) {
    stream write_stream = stropen(nemovar,"w!");
    int fd = fileno(write_stream);
    flock(fd, LOCK_EX);   // get an exclusive lock
    
    bool found_var = FALSE;
    for (i=0; i<nrows; i++) {
      s = table_row(t, i);
      if (strncmp(s,vareq,strlen(vareq))==0) {
        fprintf(write_stream,"%s=\"%s\"\n",var,value);    
        found_var = TRUE;  
      } else {
        fprintf(write_stream,"%s\n",s);
      }
    }

    if (!found_var) {
      fprintf(write_stream,"%s=\"%s\"\n",var,value);    
    }
    flock(fd, LOCK_UN);
    strclose(write_stream);
    return;
  }

  // nothing more to do yet
}
 
